File: erp/backend/routers/rh.py
from datetime import date
from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from pydantic import BaseModel
from core.deps import get_db, get_current_user
from services import rh as svc

router = APIRouter(prefix="/api/rh", tags=["rh"])


# ── Schemas ──

class FuncionarioIn(BaseModel):
    nome: str
    cpf: str
    cargo: str
    salario_base_centavos: int
    data_admissao: date
    regime: str = "clt"
    dependentes: int = 0


class FuncionarioOut(BaseModel):
    id: int; nome: str; cpf: str; cargo: str
    salario_base_centavos: int; data_admissao: date
    data_demissao: Optional[date]; regime: str; status: str
    dependentes: int
    model_config = {"from_attributes": True}


class DemitirIn(BaseModel):
    data_demissao: date


class EventoIn(BaseModel):
    funcionario_id: int
    competencia: str
    tipo: str
    valor_centavos: int
    descricao: Optional[str] = None


class EventoOut(BaseModel):
    id: int; funcionario_id: int; competencia: str
    tipo: str; valor_centavos: int; descricao: Optional[str]
    model_config = {"from_attributes": True}


class FolhaOut(BaseModel):
    id: int; competencia: str; funcionario_id: int
    salario_bruto: int; inss: int; irrf: int; fgts: int
    outros_descontos: int; salario_liquido: int; status: str
    model_config = {"from_attributes": True}


# ── Endpoints ──

@router.post("/funcionarios", response_model=FuncionarioOut, status_code=201)
def criar_funcionario(payload: FuncionarioIn, db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.criar_funcionario(
        db, payload.nome, payload.cpf, payload.cargo,
        payload.salario_base_centavos, payload.data_admissao,
        payload.regime, payload.dependentes,
    )


@router.get("/funcionarios", response_model=list[FuncionarioOut])
def listar_funcionarios(apenas_ativos: bool = Query(True),
                        db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.listar_funcionarios(db, apenas_ativos)


@router.post("/funcionarios/{func_id}/demitir", response_model=FuncionarioOut)
def demitir(func_id: int, payload: DemitirIn,
            db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.demitir_funcionario(db, func_id, payload.data_demissao)


@router.post("/eventos", response_model=EventoOut, status_code=201)
def criar_evento(payload: EventoIn, db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.criar_evento(db, payload.funcionario_id, payload.competencia,
                            payload.tipo, payload.valor_centavos, payload.descricao)


@router.get("/eventos", response_model=list[EventoOut])
def listar_eventos(competencia: str = Query(...),
                   funcionario_id: Optional[int] = Query(None),
                   db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.listar_eventos(db, competencia, funcionario_id)


@router.post("/folha/calcular", response_model=list[FolhaOut])
def calcular_folha(competencia: str = Query(...),
                   db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.calcular_folha(db, competencia)


@router.get("/folha", response_model=list[FolhaOut])
def listar_folhas(competencia: Optional[str] = Query(None),
                  db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.listar_folhas(db, competencia)


@router.get("/folha/{folha_id}/holerite")
def holerite(folha_id: int, db: Session = Depends(get_db), _=Depends(get_current_user)):
    return svc.obter_holerite(db, folha_id)

File: erp/frontend/src/api/rh.ts
import axios from 'axios'
import { formatReal } from './financeiro'
export { formatReal }

const api = axios.create({ baseURL: '/api/rh' })
api.interceptors.request.use(cfg => {
  const token = localStorage.getItem('access_token')
  if (token) cfg.headers.Authorization = `Bearer ${token}`
  return cfg
})

export interface Funcionario {
  id: number; nome: string; cpf: string; cargo: string
  salario_base_centavos: number; data_admissao: string
  data_demissao: string | null; regime: string; status: string
  dependentes: number
}

export interface EventoFolha {
  id: number; funcionario_id: number; competencia: string
  tipo: string; valor_centavos: number; descricao: string | null
}

export interface Folha {
  id: number; competencia: string; funcionario_id: number
  salario_bruto: number; inss: number; irrf: number; fgts: number
  outros_descontos: number; salario_liquido: number; status: string
}

export interface Holerite {
  funcionario: { nome: string; cpf: string; cargo: string }
  competencia: string
  salario_bruto: number; inss: number; irrf: number; fgts: number
  outros_descontos: number; salario_liquido: number
  eventos: { tipo: string; valor_centavos: number; descricao: string | null }[]
}

export const funcionariosApi = {
  listar: (apenas_ativos = true) =>
    api.get<Funcionario[]>('/funcionarios', { params: { apenas_ativos } }).then(r => r.data),
  criar: (data: Omit<Funcionario, 'id' | 'data_demissao' | 'status'>) =>
    api.post<Funcionario>('/funcionarios', data).then(r => r.data),
  demitir: (id: number, data_demissao: string) =>
    api.post<Funcionario>(`/funcionarios/${id}/demitir`, { data_demissao }).then(r => r.data),
}

export const eventosApi = {
  listar: (competencia: string, funcionario_id?: number) =>
    api.get<EventoFolha[]>('/eventos', { params: { competencia, funcionario_id } }).then(r => r.data),
  criar: (data: Omit<EventoFolha, 'id'>) =>
    api.post<EventoFolha>('/eventos', data).then(r => r.data),
}

export const folhaApi = {
  calcular: (competencia: string) =>
    api.post<Folha[]>('/folha/calcular', null, { params: { competencia } }).then(r => r.data),
  listar: (competencia?: string) =>
    api.get<Folha[]>('/folha', { params: competencia ? { competencia } : {} }).then(r => r.data),
  holerite: (id: number) =>
    api.get<Holerite>(`/folha/${id}/holerite`).then(r => r.data),
}
